### ISSUE TITLE
wishlist: be able to specify env vars in cmdline for run/exec (in presence of -e)

### ISSUE THREAD

**Issue by yarikoptic** [apptainer/singularity#2892]
E.g. as `docker run` has

```
  -e, --env list                       Set environment variables

```

For the sake of avoiding side-effects and gaining reproducibility, we tend to use `-e` (with fake/empty home provided via `-H`) quite frequently.  That obliterates all variables from the environment, which is great.  But often it is desired to set some variables for correct operation within the container environment.
The trick could be to first `env -i` for that call and then set needed variables.  Which works, BUT environment inside slightly different from what running with `-e` provides which retains e.g. `$TERM` variable.

```bash
$> env -i BU=123 singularity exec -H /tmp/fakehome -B $PWD:$PWD ./ReproNim-reproin-master-latest.simg bash -c 'export';  
declare -x BU="123"
declare -x HOME="/tmp/fakehome"
declare -x LANG="en_US.UTF-8"
declare -x LC_ALL="en_US.UTF-8"
declare -x LD_LIBRARY_PATH="/.singularity.d/libs"
declare -x ND_ENTRYPOINT="/neurodocker/startup.sh"
declare -x OLDPWD
declare -x PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
declare -x PWD="/tmp"
declare -x SHLVL="1"
declare -x SINGULARITY_APPNAME=""
declare -x SINGULARITY_CONTAINER="/tmp/ReproNim-reproin-master-latest.simg"
declare -x SINGULARITY_NAME="ReproNim-reproin-master-latest.simg"
declare -x USER_PATH=":/bin:/usr/bin:/sbin:/usr/sbin:/usr/local/bin:/usr/local/sbin"

$> singularity exec -e -H /tmp/fakehome -B $PWD:$PWD ./ReproNim-reproin-master-latest.simg bash -c 'export';      
declare -x HOME="/tmp/fakehome"
declare -x LANG="en_US.UTF-8"
declare -x LC_ALL="en_US.UTF-8"
declare -x LD_LIBRARY_PATH="/.singularity.d/libs"
declare -x ND_ENTRYPOINT="/neurodocker/startup.sh"
declare -x OLDPWD
declare -x PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
declare -x PWD="/tmp"
declare -x SHLVL="1"
declare -x SINGULARITY_APPNAME=""
declare -x SINGULARITY_CONTAINER="/tmp/ReproNim-reproin-master-latest.simg"
declare -x SINGULARITY_NAME="ReproNim-reproin-master-latest.simg"
declare -x TERM="rxvt-unicode-256color"

```
  So I thought it would have been nice for consistency to be able to specify those environment variables explicitly to `run` or `exec` call

**Comment by bauerm97** [comment]
Hey @yarikoptic, I could have sworn I replied to this two days ago when I added the label to it. I don't see any reason why we can't add a flag supporting this. Here's my initial proposal for what we can do here:

`--set-env <string list>`: Comma separated list of `Key=Value` pairs, which are guaranteed to be set in the final environment in the container

**Comment by bauerm97** [comment]
Let me know if that would adequately address your request and we can try to get the ball rolling with a developer on the Sylabs team or can help you contribute it as a patch if you're interested

**Comment by yarikoptic** [comment]
Sounds good, besides the "comma separated" aspect. I don't see why comma can't be used in a value for a variable, eg we do have a few of those at least in pymvpa and datalad, eg to list  metrics to log.

**Comment by yarikoptic** [comment]
Re patch from yours truly, not sure if I would pull it off ATM.

**Comment by bauerm97** [comment]
>Sounds good, besides the "comma separated" aspect. I don't see why comma can't be used in a value for a variable, eg we do have a few of those at least in pymvpa and datalad, eg to list metrics to log.

This is a very good point. By default, the CLI flags framework we use separates environment variable values by comma. So `--flag val1,val2,val3` is parsed by the pflag framework into `[]string{val1,val2,val3}`. I'm not sure how we'll be able to tackle that, but I'm happy to take a look and investigate further.

**Comment by bauerm97** [comment]
Actually, after a bit of reading into `pflag`, it looks like we'll be able to use the [StringArray](https://godoc.org/github.com/spf13/pflag#StringArray) class of flag values to not attempt to parse commas as separator tokens. The usage of this would essentially be:

`--set-env key1=val1 --set-env key2=val2`

**Comment by yarikoptic** [comment]
Yeap, that would be great! Thanks in advance!

**Comment by yarikoptic** [comment]
another finding of today which might be implemented along: If no `=value` provided for a variable - pass that variable value into container (in case of `-e`).  So, in my use case I would have used `--set-env USER` since USER is cleared out, not automagically set by the bash inside, and seems to be needed by some tools, and probably indeed worthwhile filtering out by default in case of `-e`:
```shell
$> echo $USER
yoh

$> singularity exec -e ./neurodebian-latest.img bash -c 'export'
declare -x HOME="/home/yoh"
declare -x LANG="C"
declare -x LD_LIBRARY_PATH="/.singularity.d/libs"
declare -x OLDPWD
declare -x PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
declare -x PWD="/home/yoh/repronim/containers/singularity"
declare -x SHLVL="1"
declare -x SINGULARITY_APPNAME=""
declare -x SINGULARITY_CONTAINER="/home/yoh/repronim/containers/singularity/neurodebian-latest.img"
declare -x SINGULARITY_NAME="neurodebian-latest.img"
declare -x TERM="rxvt-unicode-256color"
```

**Comment by yarikoptic** [comment]
a "finding" that there is already a way!  Just export desired variables with `SINGULARITYENV_` prefix:

```shell
$> SINGULARITYENV_PASSME=1 /home/yoh/proj/repronim/containers/scripts/singularity_cmd exec ~/proj/repronim/containers/images/repronim/repronim-reproin--0.5.4.sing bash -c 'export'
declare -x HOME="/home/yoh/proj/repronim/containers/binds/HOME"
declare -x LANG="en_US.UTF-8"
declare -x LC_ALL="en_US.UTF-8"
declare -x LD_LIBRARY_PATH="/.singularity.d/libs"
declare -x ND_ENTRYPOINT="/neurodocker/startup.sh"
declare -x OLDPWD
declare -x PASSME="1"
declare -x PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
declare -x PWD="/home/yoh/deb/gits/pkg-exppsy/singularity-container"
declare -x SHLVL="1"
declare -x SINGULARITY_APPNAME=""
declare -x SINGULARITY_CONTAINER="/home/yoh/proj/repronim/containers/images/repronim/repronim-reproin--0.5.4.sing"
declare -x SINGULARITY_NAME="repronim-reproin--0.5.4.sing"
declare -x TERM="rxvt-unicode-256color"
```
(that helper script runs singularity with -e etc so it is all working out as desired).  Feel free to close I guess ;)

**Comment by GodloveD** [comment]
Glad you found a solution (and recorded it for others too)!  Thanks @yarikoptic

**Comment by oesteban** [comment]
I'd like to reopen this issue. This works well for variables undefined in the container, but can't overwrite existing ones (which is quite limiting):

TEMPLATEFLOW_HOME is pre-set to /opt/templateflow in the container.
```
$ SINGULARITYENV_TEMPLATEFLOW_HOME=$HOME/.cache/templateflow singularity exec /home/groups/russpold/singularity_images/poldracklab_fmriprep_1.4.1rc2-2019-06-07-03d2c6cafaf5.simg env | grep TEMPLATEFLOW
TEMPLATEFLOW_HOME=/opt/templateflow
$ SINGULARITYENV_TEMPLATEFLOWHOME=$HOME/.cache/templateflow singularity exec /home/groups/russpold/singularity_images/poldracklab_fmriprep_1.4.1rc2-2019-06-07-03d2c6cafaf5.simg env | grep TEMPLATEFLOW
TEMPLATEFLOW_HOME=/opt/templateflow
TEMPLATEFLOWHOME=/home/users/oesteban/.cache/templateflow
```

**Comment by dtrudg** [comment]
This is included in the env var rework in #5028
